relay-server/src/services/projects/cache/state.rs

```diff
@@ -0,0 +1,774 @@
+///
+/// This state is interior mutable and allows updates to the project.
+#[derive(Debug, Default, Clone)]
+struct SharedProjectState(Arc<ArcSwap<SharedProjectStateInner>>);
+
+impl SharedProjectState {
+    /// Updates the project state.
+    fn set_project_state(&self, state: ProjectState) {
+        let prev = self.0.rcu(|stored| SharedProjectStateInner {
+            state: state.clone(),
+            rate_limits: Arc::clone(&stored.rate_limits),
+            reservoir_counters: Arc::clone(&stored.reservoir_counters),
+        });
+
+        // Try clean expired reservoir counters.
+        //
+        // We do it after the `rcu`, to not re-run this more often than necessary.
+        if let Some(state) = state.enabled() {
+            let config = state.config.sampling.as_ref();
+            if let Some(config) = config.and_then(|eb| eb.as_ref().ok()) {
+                // We can safely use previous here, the `rcu` just replaced the state, the
+                // reservoir counters did not change.
+                //
+                // `try_lock` to not potentially block, it's a best effort cleanup.
+                //
+                // TODO: Remove the lock, we already have interior mutability with the `ArcSwap`
+                // and the counters themselves can be atomics.
+                if let Ok(mut counters) = prev.reservoir_counters.try_lock() {
+                    counters.retain(|key, _| config.rules.iter().any(|rule| rule.id == *key));
+                }
+            }
+        }
+    }
+
+    /// Extracts and clones the revision from the contained project state.
+    fn revision(&self) -> Revision {
+        self.0.as_ref().load().state.revision().clone()
+    }
+
+    /// Transforms this interior mutable handle to an immutable [`SharedProject`].
+    fn to_shared_project(&self) -> SharedProject {
+        SharedProject(self.0.as_ref().load_full())
+    }
+}
+
+/// The data contained in a [`SharedProjectState`].
+///
+/// All fields must be cheap to clone and are ideally just a single `Arc`.
+/// Partial updates to [`SharedProjectState`], are performed using `rcu` cloning all fields.
+#[derive(Debug, Default)]
+struct SharedProjectStateInner {
+    state: ProjectState,
+    rate_limits: Arc<CachedRateLimits>,
+    reservoir_counters: ReservoirCounters,
+}
+
+/// Current fetch state for a project.
+#[derive(Debug)]
+enum FetchState {
+    /// There is a fetch currently in progress.
+    InProgress,
+    /// A successful fetch is pending.
+    ///
+    /// This state is essentially only the initial state, a project
+    /// for the most part should always have a fetch in progress or be
+    /// in the non-pending state.
+    Pending {
+        /// Time when the next fetch should be attempted.
+        ///
+        /// `None` means soon as possible.
+        next_fetch_attempt: Option<Instant>,
+    },
+    /// There was a successful non-pending fetch,
+    NonPending {
```

Comment:
I settled on `NonPending`, because this represents all states of the `ProjectState` which are not pending (`Disabled` and `Enabled`).

I like `Complete`, thanks!